### PR TITLE
add Firefox support

### PR DESCRIPTION
**Pull request by YusukeIwaki**
refs #117 

We can support Firefox by migrating `FirefoxLauncher` class from [here](https://github.com/puppeteer/puppeteer/blob/main/src/node/Launcher.ts) into Dart code.
(I actually developed it in my personal project [puppeteer-ruby](https://github.com/YusukeIwaki/puppeteer-ruby/pull/21) )

puppeteer-dart currently doesn't have `Launcher` and Puppeteer class has such functionality, so I tried to create another puppeteer class as `PuppeteerFirefox`

Known issues:

* Firefox often returns imcomplete payloads (some fields are `null` or `{}` or undefined...). Current `protocol` classes are assumed to be used with Chromium, and often throw cast errors with Firefox.
* ~~Firefox nightly is distributed with variours formats (.zip for Win64, .dmg for macOS, .tar.bz2 for Linux), and it is a little hard to support extracting all of them. (I always use Docker, and actually I don't use it... 😅 ) So in this implementation, `executablePath` is assumed to be required. (Downloader for Firefox is not implemented)~~ IMPLEMENTED

### example

```dart
Future<void> main() async {
  Logger.root.level = Level.ALL;
  Logger.root.onRecord.listen(print);

  final browser = await puppeteerFirefox.launch(
    headless: false,
    executablePath: '/Applications/Firefox Nightly.app/Contents/MacOS/firefox',
  );
  final myPage = await browser.newPage();

  // Go to a page and wait to be fully loaded
  await myPage.goto('https://www.google.com');

  // Do something... See other examples
  await myPage.click('input[name="q"]');
  await myPage.keyboard.type('puppeteer');

  await Future.wait([
    myPage.waitForNavigation(),
    myPage.keyboard.press(Key.enter),
  ]);

  await File('_google-puppeteer.png').writeAsBytes(await myPage.screenshot());

  await myPage.goto('https://detectmybrowser.com/');

  await File('_detect-my-browser.png').writeAsBytes(await myPage.screenshot());

  // Gracefully close the browser's process
  await browser.close();
}
```

_google-puppeteer.png:

google-puppeteer

_detect-my-browser.png

detect-my-browser

### CODE REVIEW

**Comment by xvrh** [comment]
Hi @YusukeIwaki ,
Thank you for sharing your work. This is awesome.

The approach looks good to me. And I think it's ok to not have the download/extract part for Firefox.

The main issue is the incomplete/bad payload for firefox. It will require to modify the code generator to be more lenient.
It would also be good to run all the tests with Firefox.

Good job 👍

File: example/example.dart [hunk 1]
```diff
@@ -8,11 +9,11 @@ void main() async {
   var myPage = await browser.newPage();
 
   // Go to a page and wait to be fully loaded
-  await myPage.goto('https://www.github.com', wait: Until.networkIdle);
+  await myPage.goto('https://www.github.com', wait: Until.domContentLoaded);
 
   // Do something... See other examples
   await myPage.screenshot();
-  await myPage.pdf();
+  await myPage.pdf(output: File('github_my_page.pdf').openWrite());
```
Comment (YusukeIwaki) [inline example/example.dart:16]: An argument error occurs on Firefox without this parameter... :(

File: example/keyboard.dart [hunk 2]
```diff
@@ -24,7 +24,12 @@ void main() async {
 
   await page.bringToFront();
 
-  await page.keyboard.type('éèà Hello');
+  if (isPuppeteerFirefox) {
+    // Firefox cannot input 'è' with Input.insertText
+    await page.keyboard.type('Hello World!');
+  } else {
+    await page.keyboard.type('éèà Hello');
```
Comment (YusukeIwaki) [inline example/keyboard.dart:31]: Currenrly, Firefox doesn't accept any emoji.

File: example/search.dart [hunk 3]
```diff
@@ -16,7 +16,7 @@ void main() async {
   await page.click(allResultsSelector);
 
   // Wait for the results page to load and display the results.
-  const resultsSelector = '.gsc-results .gsc-thumbnail-inside a.gs-title';
+  const resultsSelector = '.gsc-result a.gs-title';
```
Comment (YusukeIwaki) [inline example/search.dart:19]: This test sometimes fails also on Chromium.
It seems that the google result page sometimes returns the page with different DOM structure.

**Comment by YusukeIwaki** [comment]
@xvrh Could you take a look at this PR?

**Comment by xvrh** [comment]
Hi @YusukeIwaki 
Sorry for the delay.
Thanks a lot for the hard work to find the small differences between chrome and Firefox.

@kendalsickels-wf maybe you are interested in trying this, since you opened #117 

We cannot merge it as it because some of the files that you changed are generated from the protocol schema. We will need to  change the generator to include your changes.

**Comment by YusukeIwaki** [comment]
Very sorry, I couldn't afford to fix the problems...

Now I already use Playwright for Firefox automation because the CDP implementation of Firefox is still really limited...
Please let me close this PR. I hope another Firefox user would contribute the Firefox support.

**Pull request closed without merging**